feat(hardware): ot3: add pipette querying

sfoster1 opened the pull request:
With a new GetPipetteInfo message that can send back the name, model, and serial (similar semantics as in shared data) we can now hook up the pipette identification machinery to work the same as in the OT-2.

sfoster1 committed:
feat(hardware): ot3: add pipette info messages

These messages can tell the host what kind of pipette is attached.

sfoster1 committed:
feat(hardware): ot3: add tool info detector

We currently can detect what the head says is attached to it, but that
doesn't give us a ton of details. Now that we can query pipette
information from the pipettes directly, we can chain
- Asking what the head has
- Asking pipettes that the head says should be attached what they are

While properly only waiting for the ones that are attached to answer.

We will want to make a reactive version of this when we do the hardware
controller's attached tools being reactive, but to avoid extra churn the
first time through this is just a oneshot.

sfoster1 committed:
refactor(api): ot3: detect attached tools

Based on the new tool info detector, we can actually implement
get_attached_instruments() by querying the canbus. In many ways, this
duplicates the functionality of probe_network(); this now moves into the
realm of "debug tool" and perhaps should be removed.

amitlissack commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -272,18 +274,42 @@ async def get_attached_instruments(
+
+        current_tools = dict(_generate_attached_instrs(attached))
+        self._present_nodes -= set(
+            (
```
Nit: Extraneous parentheses?

amitlissack commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -469,6 +495,26 @@ def home_position() -> OT3AxisMap[float]:
+        Unlike probe_network, this always waits for the nodes that must be present for
+        a working machine, and no more.
+        """
+        core_nodes = set([NodeId.gantry_x, NodeId.gantry_y, NodeId.head])
```
This is really important logic in this method.

A static function that gets  `self._present_nodes` what is returned by `self.probe` and returns what should be assigned to `self._present_nodes` (or even mutates it) would be a lot easier to test.

sfoster1 replied:
yeah good point

amitlissack commented on hardware/tests/conftest.py:
```diff
@@ -39,3 +41,75 @@ def mock_messenger(can_message_notifier: MockCanMessageNotifier) -> AsyncMock:
     return mock
+
+
+class CanLoopback:
```
A+

amitlissack reviewed (commented):
Really great. One little question.

sfoster1 committed:
refactor(api): factor out probe filtering

amitlissack reviewed (approved):
Aces

sfoster1 committed:
refactor(api): ot3: load gantry load from instrs

This will need to be updated to properly decide between low and high
throughput instruments and account for the gripper, but for now it
should work.

Laura-Danielle reviewed (approved):
looks good to me
